### ISSUE TITLE
T480 compatibility

### ISSUE THREAD

**Issue by EETagent** [MSzturc/ThinkpadAssistant#8]
Hello, 
I would like to ask you for little assistance with the T480 keyboard mapping to make it work with the Thinkpad Assistant. I'm quite a noob when it comes to the ACPI patches.

[DSDT.zip](https://github.com/MSzturc/ThinkpadAssistant/files/4662380/DSDT.zip)
[config.zip](https://github.com/MSzturc/ThinkpadAssistant/files/4662383/config.zip)
[SSDT-BKeyQ14Q15-TP-LPCB.zip](https://github.com/MSzturc/ThinkpadAssistant/files/4662385/SSDT-BKeyQ14Q15-TP-LPCB.zip)

Keys currently works as normal FN keys with backlight control using BKeyQ14Q15-TP-LPCB

Keyboard part (?) from my DSDT
```

Method (XQ14, 0, NotSerialized)
            {
                If (\_SB.PCI0.LPCB.EC.HKEY.MHKK (0x01, 0x8000))
                {
                    \_SB.PCI0.LPCB.EC.HKEY.MHKQ (0x1010)
                }

                If (\VIGD)
                {
                    Notify (\_SB.PCI0.GFX0.DD1F, 0x86) // Device-Specific
                }
            }

            Method (XQ15, 0, NotSerialized)
            {
                If (\_SB.PCI0.LPCB.EC.HKEY.MHKK (0x01, 0x00010000))
                {
                    \_SB.PCI0.LPCB.EC.HKEY.MHKQ (0x1011)
                }

                If (\VIGD)
                {
                    Notify (\_SB.PCI0.GFX0.DD1F, 0x87) // Device-Specific
                }

                Return (Zero)
            }

            Method (BRNS, 0, NotSerialized)
            {
                Local0 = (\BRLV + 0x02)
                Local3 = \BNTN
                If (\_SB.PCI0.GFX0.DRDY)
                {
                    Local2 = DerefOf (DerefOf (BRTB [Local3]) [Local0])
                    \_SB.PCI0.GFX0.AINT (0x01, Local2)
                }
            }

            Method (BFRQ, 0, NotSerialized)
            {
                Local0 = 0x80000100
                Local1 = DerefOf (DerefOf (BRTB [\BNTN]) [0x13])
                Local0 |= (Local1 << 0x09)
                Local1 = DerefOf (DerefOf (BRTB [\BNTN]) [0x15])
                Local0 |= Local1
                Return (Local0)
            }
        }

        Scope (\_SB.PCI0.LPCB.EC)
        {
            Name (BDEV, 0xFF)
            Name (BSTS, 0x00)
            Name (BHKE, 0x00)
            Method (_Q2C, 0, NotSerialized)  // _Qxx: EC Query, xx=0x00-0xFF
            {
                If ((BSTS == 0x00))
                {
                    BDEV = BGID (0x00)
                    NBRE (BDEV)
                }
            }

            Method (_Q2D, 0, NotSerialized)  // _Qxx: EC Query, xx=0x00-0xFF
            {
                BDEV = BGID (0x00)
                NBIN (BDEV)
            }

            Method (_Q38, 0, NotSerialized)  // _Qxx: EC Query, xx=0x00-0xFF
            {
                Local0 = BGID (0x00)
                If ((Local0 == 0x0F))
                {
                    BDIS ()
                    \BHDP (0x01, 0x00)
                    NBEJ (BDEV)
                    BDEV = Local0
                    If ((\BIDE == 0x03))
                    {
                        \_SB.PCI0.SAT0.PRIM.GTME = 0x00
                        \_SB.PCI0.SAT0.SCND.GTME = 0x00
                    }
                }
                ElseIf (HPBU){}
                Else
                {
                    BDEV = Local0
                    NBIN (Local0)
                }
            }

```
Thanks for help

**Comment by MSzturc** [comment]
Try this:
- Copy SSDT-KBRD.aml into EFI/OC/ACPI
- Activate it by patching SSDT in config.plist. Here is a List of all patches you need:

![image](https://user-images.githubusercontent.com/4275753/82585393-397b4100-9b96-11ea-980b-c15d46250011.png)

**Comment by EETagent** [comment]
@MSzturc  Thanks for your response

Copying SSDT-KBRD would not work, because my DSDT use LPCB instead of LPC. When I try to rename LPC to LPCB I get 6 six weird errors. What could went wrong? 

![Snímek obrazovky 2020-05-22 v 16 26 53](https://user-images.githubusercontent.com/20557318/82678088-4dcc4600-9c49-11ea-80be-dde3615b82ed.png)

**Comment by EETagent** [comment]
@MSzturc  Thanks for response

I added those renames and KBRD.aml to my EFI and it resulted in Kernel Panic. My DSDT uses LPCB instead of LPC so I replaced that. I've got also compile errors in the highlighted section, removing __ after LPCB solved that
![Snímek obrazovky 2020-05-22 v 16 26 53](https://user-images.githubusercontent.com/20557318/82681203-bb7a7100-9c4d-11ea-8712-3114212fb9cd.png)

[DSDT.dsl.zip](https://github.com/MSzturc/ThinkpadAssistant/files/4668851/DSDT.dsl.zip)
[config.plist.zip](https://github.com/MSzturc/ThinkpadAssistant/files/4668852/config.plist.zip)
[SSDT-KBRD.aml.zip](https://github.com/MSzturc/ThinkpadAssistant/files/4668853/SSDT-KBRD.aml.zip)

**Comment by MSzturc** [comment]
Thx for the working configurations. Added them as samples for users in the future

**Comment by EETagent** [comment]
@MSzturc  I apologize for the misunderstanding. I probably wrote it wrong. This config does not work. It resulted in kernel panic like I wrote above.

**Comment by MSzturc** [comment]
There has to be a typo in your ssdt but i see it right now. I would suggest that you try the ssdt from @xma https://github.com/xma/T480-Clover/blob/master/CLOVER/ACPI/patched/SSDT-KBD.dsl He maps only two keys right now but it if it works it could be a nice starting point to go further.

**Comment by EETagent** [comment]
@MSzturc  Thanks, nice find. It's working well now. As soon as I have time I will try to add Bluetooth and screen extending/mirroring support.

**Comment by xma** [comment]
Yeah it's working fine for clover, would you mind adding it in README.md ? thank you

**Comment by xma** [comment]
Thanks for the link. I've added missing keys onto T480 - layout is same as X1C6, its work really well !

**Comment by MSzturc** [comment]
Updated Readme with new samples

@EETagent please check if all Function Keys are working

**Comment by EETagent** [comment]
@xma  @MSzturc  Thanks for help

- [x] Brightness
- [x] Microphone
- [x] Screen mirroring
- [ ]  Screen extending - I am unable to switch back to extending after changing to mirroring mode.
- [x] Wifi. Probably. I do not have working Wi-Fi card
- [x] Bluetooth 
- [x] Star key

**Comment by MSzturc** [comment]
Are you able to switch modes by Airplay Sidecar Menu Bar Entry?

![image](https://user-images.githubusercontent.com/4275753/82843032-19bc8380-9edc-11ea-96eb-10d7e80ee146.png)

**Comment by EETagent** [comment]
@MSzturc 

No. But I'm able to change mode using display settings
![Snímek obrazovky 2020-05-25 v 23 42 44](https://user-images.githubusercontent.com/20557318/82844675-a584de80-9ee1-11ea-92ad-cb40cd7fb0f2.png)

**Comment by MSzturc** [comment]
I don't have a multi monitor setup here to test it. I would guess that this line causes the problem:

![image](https://user-images.githubusercontent.com/4275753/82883170-60dd5f80-9f42-11ea-8dfd-97a139027303.png)

Are you able to start Thinkpad Assistant out of XCode and try it from there? I put assertion inside the Code. You should see Log in the Console if an Error happens

**Comment by EETagent** [comment]
@MSzturc  Here is the log

```
2020-05-27 12:09:25.052942+0200 ThinkpadAssistant[1501:36218] Metal API Validation Enabled
2020-05-27 12:09:33.882870+0200 ThinkpadAssistant[1501:36218] [plugin] AddInstanceForFactory: No factory registered for id <CFUUID 0x6000002560e0> F8BB1C28-BAE8-11D6-9C31-00039315CD46
2020-05-27 12:09:33.924972+0200 ThinkpadAssistant[1501:36218]  HALC_ShellDriverPlugIn::Open: Can't get a pointer to the Open routine
2020-05-27 12:09:33.925543+0200 ThinkpadAssistant[1501:36218]  HALC_ShellDriverPlugIn::Open: Can't get a pointer to the Open routine
2020-05-27 12:09:33.925989+0200 ThinkpadAssistant[1501:36218]  HALC_ShellDriverPlugIn::Open: Can't get a pointer to the Open routine
2020-05-27 12:10:07.292030+0200 ThinkpadAssistant[1501:36678]  HALC_ShellDriverPlugIn::Open: Can't get a pointer to the Open routine

```

When mirroring and trying to get back to the extending, no log is given. Logs above appears when using Microphone switch or monitor mirroring for the first time.

**Comment by MSzturc** [comment]
Hmm that's a strange behaviour.

The Code of the DisplayManager is based on this tool: https://github.com/fcanas/mirror-displays Basically it's a enhanced port of it to Swift. Can you check if you are able to switch between mirroring and extending with this tool. I want to make sure that it behaves diffrent.

**Comment by EETagent** [comment]
@MSzturc  Downloaded mirror executable, and it works. I am able to switch between mirroring and extending.

**Comment by EETagent** [comment]
No logs were displayed this time. Only Metal Validation appeared when I clicked on app icon. The application is compiled correctly. It works otherwise normally.

![Snímek obrazovky 2020-05-27 v 12 34 31](https://user-images.githubusercontent.com/20557318/83009309-db77af00-a016-11ea-8f44-8c1110e8d4c4.png)

**Comment by MSzturc** [comment]
fixed with 1ae384ea2e6b0a3f9a1d6cc0cd7640eeadb89370

**Comment by EETagent** [comment]
> fixed with [1ae384e](https://github.com/MSzturc/ThinkpadAssistant/commit/1ae384ea2e6b0a3f9a1d6cc0cd7640eeadb89370)
> 
> I will release a new version as soon as i finished the keyboard backlight feature

I can confirm that this is fixed. 
![Snímek obrazovky 2020-05-27 v 14 01 17](https://user-images.githubusercontent.com/20557318/83016633-90fc2f80-a022-11ea-80d6-734ed4fbbae5.png)

**Comment by EETagent** [comment]
@MSzturc  

Thanks to @xma I realised my previous mistakes with KBRD.aml and was successfuly able to add your recent ACPI patch for backlight control. Is there currently some way how to test if it works?

**Comment by MSzturc** [comment]
I've commited samples for T460 and the Backlight functionality. Have fun testing ;-)

**Comment by EETagent** [comment]
@MSzturc  It works great. 

Constructive criticism: Personally, I'm not a fan of current icons. I think something like that would be much better


![](https://user-images.githubusercontent.com/30008333/82955786-cddb0e80-9faf-11ea-8e5c-68b900da7cf7.png)

**Comment by MSzturc** [comment]
Closed since it's working with v1.7.0. Moved the discussion about the HUD Layout to a new central topic #16
